src/audio_core/hle/mediandk_decoder.cpp

```diff
@@ -0,0 +1,214 @@
+
+#include <media/NdkMediaCodec.h>
+#include <media/NdkMediaError.h>
+#include <media/NdkMediaFormat.h>
+
+#include <vector>
+
+#include "audio_core/hle/mediandk_decoder.h"
+
+namespace AudioCore::HLE {
+
+class MediaNDKDecoder::Impl {
+public:
+    explicit Impl(Memory::MemorySystem& memory);
+    ~Impl();
+    std::optional<BinaryResponse> ProcessRequest(const BinaryRequest& request);
+
+    bool SetMediaType(u32 num_channels, u32 sample_rate);
+
+private:
+    std::optional<BinaryResponse> Initalize(const BinaryRequest& request);
+    std::optional<BinaryResponse> Decode(const BinaryRequest& request);
+
+    Memory::MemorySystem& mMemory;
+    AMediaCodec* mDecoder = nullptr;
+    u32 mNumChannels = 0;
+    u32 mSampleRate = 0;
+};
+
+MediaNDKDecoder::Impl::Impl(Memory::MemorySystem& memory) : mMemory(memory) {
+    SetMediaType(2, 48000);
+}
+
+MediaNDKDecoder::Impl::~Impl() {
+    if (mDecoder) {
+        AMediaCodec_delete(mDecoder);
```

Comment:
Just a nitpick, the recommended way to clean up the decoder is:

```suggestion
        AMediaCodec_stop(mDecoder);
        AMediaCodec_delete(mDecoder);
```

It will be probably fine with the current code, but certain Android vendors may register their AAC decoder which may not be happy with only deleting the decoder instance without stopping it first. (They probably won't expect anyone to use NDK version of the API)

An even better way would be taking @wwylele 's suggestion and create a smart pointer to wrap the instance. And you might want to write a cleanup function with something like:

```c++
struct AMediaCodecRelease {
    void operator()(AMediaCodec* codec) const {
        AMediaCodec_stop(codec);
        AMediaCodec_delete(codec);
    };
};
```